src/client/datascience/ipywidgets/notebookIPyWidgetCoordinator.ts

```diff
@@ -1,74 +1,209 @@
 // Licensed under the MIT License.
 import { inject, injectable } from 'inversify';
-import { NotebookCommunication, NotebookDocument, CancellationToken, Disposable } from 'vscode';
-import { IVSCodeNotebook } from '../../common/application/types';
+import {
+    NotebookDocument,
+    CancellationToken,
+    Disposable,
+    NotebookEditor,
+    window,
+    Uri,
+    EventEmitter,
+    notebook,
+    CancellationTokenSource
+} from 'vscode';
 import { Cancellation } from '../../common/cancellation';
+import { disposeAllDisposables } from '../../common/helpers';
+import { traceInfo } from '../../common/logger';
+import { IAsyncDisposableRegistry, IDisposable, IDisposableRegistry } from '../../common/types';
 import { createDeferred } from '../../common/utils/async';
+import { noop } from '../../common/utils/misc';
 import { IServiceContainer } from '../../ioc/types';
+import { ConsoleForegroundColors } from '../../logging/_global';
 import { InteractiveWindowMessages, IPyWidgetMessages } from '../interactive-common/interactiveWindowTypes';
+import { INotebookCommunication } from '../notebook/types';
+import { VSCodeNotebookController } from '../notebook/vscodeNotebookController';
 import { CommonMessageCoordinator } from './commonMessageCoordinator';
 
+class NotebookCommunication implements INotebookCommunication, IDisposable {
+    private eventHandlerListening?: boolean;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    private pendingMessages: any[] = [];
+    private readonly disposables: IDisposable[] = [];
+    private controllerMessageHandler?: IDisposable;
+    private controller!: VSCodeNotebookController;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    private readonly _onDidReceiveMessage = new EventEmitter<any>();
+    constructor(public readonly editor: NotebookEditor, controller: VSCodeNotebookController) {
+        this.changeController(controller);
+    }
+    public changeController(controller: VSCodeNotebookController) {
+        if (this.controller === controller) {
+            return;
+        }
+        this.controllerMessageHandler?.dispose();
+        this.controller = controller;
+        this.controllerMessageHandler = controller.onDidReceiveMessage((e) => {
+            // Handle messages from this only if its still the active controller.
+            if (e.editor === this.editor && this.controller === controller) {
+                // If the listeners haven't been hooked up, then dont fire the event (nothing listening).
+                // Instead buffer the messages and fire the events later.
+                if (this.eventHandlerListening) {
+                    this.sendPendingMessages();
+                    this._onDidReceiveMessage.fire(e.message);
+                } else {
+                    this.pendingMessages.push(e.message);
+                }
+            }
+        });
+    }
+    public dispose() {
+        disposeAllDisposables(this.disposables);
```

Comment:
Do we also need a dispose of controllerMessageHandler here? I don't think that it's added to disposables, only disposed on changeController.

Reply:
Will check

Reply:
Yes, I missed that one, thanks